dali/operators/decoder/nvjpeg/nvjpeg_decoder_decoupled_api.h

```diff
@@ -521,100 +527,124 @@ class nvJPEGDecoder : public Operator<MixedBackend>, CachedDecoderImpl {
+        nvjpegChromaSubsampling_t subsampling;
+        nvjpegStatus_t ret = nvjpegGetImageInfo(handle_, input_data, in_size, &c,
+                                                &subsampling, widths, heights);
+        bool hw_decode = false;
+        bool nvjpeg_decode = (ret == NVJPEG_STATUS_SUCCESS);
+
+        auto crop_generator = GetCropWindowGenerator(i);
+        if (nvjpeg_decode) {
+          data.shape = {heights[0], widths[0], c};
+          data.subsampling = subsampling;
+        } else if (crop_generator || !ParseNvjpeg2k(data,
+                                                    span<const uint8_t>(input_data, in_size))) {
+          try {
+            data.method = DecodeMethod::Host;
+            auto image = ImageFactory::CreateImage(input_data, in_size, output_image_type_);
+            data.shape = image->PeekShape();
+          } catch (const std::runtime_error &e) {
+            DALI_FAIL(e.what() + ". File: " + data.file_name);
           }
         }
-#endif
-        if (hw_decode) {
-          data.method = DecodeMethod::NvjpegHw;
-          samples_hw_batched_.push_back(&data);
+        data.req_nchannels = NumberOfChannels(output_image_type_, data.shape[2]);
+
+        if (crop_generator) {
+          TensorShape<> dims{data.shape[0], data.shape[1]};
+          data.roi = crop_generator(dims, "HW");
+          DALI_ENFORCE(data.roi.IsInRange(dims));
+          output_shape_.set_tensor_shape(
+            i, {data.roi.shape[0], data.roi.shape[1], data.req_nchannels});
+          NVJPEG_CALL(nvjpegDecodeParamsSetROI(data.params, data.roi.anchor[1], data.roi.anchor[0],
+                                              data.roi.shape[1], data.roi.shape[0]));
         } else {
-          data.method = DecodeMethod::NvjpegCuda;
-          samples_single_.push_back(&data);
+          output_shape_.set_tensor_shape(i, {data.shape[0], data.shape[1], data.req_nchannels});
+          NVJPEG_CALL(nvjpegDecodeParamsSetROI(data.params, 0, 0, -1, -1));
         }
-      }
 
-      data.is_progressive = IsProgressiveJPEG(input_data, in_size);
-      if (data.method == DecodeMethod::NvjpegCuda) {
-        int64_t sz = data.roi
-          ? data.roi.shape[1] * (data.roi.anchor[0] + data.roi.shape[0])
-          : data.shape[0] * data.shape[1];
-        if (sz > hybrid_huffman_threshold_ && !data.is_progressive) {
-          data.selected_decoder = &data.decoders[NVJPEG_BACKEND_GPU_HYBRID];
-        } else {
-          data.selected_decoder = &data.decoders[NVJPEG_BACKEND_HYBRID];
+        // only when we have ROI info check if nvjpegDecodeBatchedSupportedEx supports it
+        if (nvjpeg_decode) {
+  #if IS_HW_DECODER_COMPATIBLE
+          if (state_hw_batched_ != nullptr) {
+            // in some cases hybrid decoder can handle the image but HW decoder can't, we should not
+            // error in that case
+            ret = nvjpegJpegStreamParse(handle_, input_data, in_size, 0, 0,
```

Comment:
These stream handles are per-sample and don't appear to be special or different from those used by the hybrid decoder. Can we reuse these already parsed streams for the regular decoder? In this case we'd need an array of booleans which we would zero-fill at the beginning of this function and here we would store:
`jpeg_stream_ready_[i] = (ret == NVJPEG_STATUS_SUCCESS)`.
Then in the ProcessSample we could just use the already parsed stream.

Reply:
OK, now they're per-thread - nvm.